pkg/activator/handler/revision_handler.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handler
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"knative.dev/pkg/network"
+	"knative.dev/serving/pkg/activator"
+)
+
+// RevisionHandler is a handler that decode the revision name and namespace from the Host header.
+type RevisionHandler struct {
+	NextHandler http.Handler
+	Domain      string
+}
+
+// NewRevisionHandler creates a handler that if the headers aren't explicitly specified,
+// then decode the revision name and namespace from the Host header and set it to the header.
+func NewRevisionHandler(h http.Handler) http.Handler {
+	return &RevisionHandler{
+		NextHandler: h,
+		Domain:      network.GetClusterDomainName(),
+	}
+}
+
+func (h *RevisionHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	namespace := r.Header.Get(activator.RevisionHeaderNamespace)
+	name := r.Header.Get(activator.RevisionHeaderName)
+	if name != "" && namespace != "" {
+		h.NextHandler.ServeHTTP(rw, r)
+		return
+	}
+
+	if n, ns, ok := getNameAndNamespace(r.Host, h.Domain); ok {
+		r.Header.Set(activator.RevisionHeaderNamespace, ns)
+		r.Header.Set(activator.RevisionHeaderName, n)
+		h.NextHandler.ServeHTTP(rw, r)
+		return
+	}
+
+	msg := fmt.Sprintf(`One of the header and the host must be set correctly, Header (%s=%q, %s=%q), Host=%q`,
+		activator.RevisionHeaderName, name,
+		activator.RevisionHeaderNamespace, namespace,
+		r.Host)
+	http.Error(rw, msg, http.StatusNotFound)
+}
+
+func getNameAndNamespace(host string, domain string) (name string, namespaces string, ok bool) {
+	parts := strings.SplitN(host, ".", 4)
+	switch len(parts) {
+	case 2:
+	case 3:
```

Comment:
I'm still not sure that we actually want to introduce this new behaviour to accept more types of host header. Could we remove this file for now, and the new switch statement, and open another issue/PR to discuss that? Let's keep this PR focused only on changing request_log to use the Revision set in the context by context_handler.